addon/mixins/spread.js

```diff
@@ -1,73 +1,196 @@
+    // Get or create the array of spread listeners on the source object and add this object
+    const spreadListeners = get(sourceObject, `${sourceProperty}._spreadListeners`)
+    if (isNone(spreadListeners)) {
+      get(sourceObject, sourceProperty).set('_spreadListeners', [
+        {
+          targetObject: this,
+          targetProperty: spreadProperty
         }
+      ])
+    } else {
+      spreadListeners.push({
+        targetObject: this,
+        targetProperty: spreadProperty
+      })
+    }
 
-        defineProperty(get(spreadContext, spreadSource), '_spreadListeners', [])
-        defineProperty(get(spreadContext, spreadSource), 'setUnknownProperty', undefined,
-          function (key, value) {
-            this[key] = value
-            this._spreadListeners.forEach(listener => {
-              defineProperty(listener, key, readOnly(`${spreadProperty}.${key}`))
-            })
-            spreadContext.get(spreadSource).notifyPropertyChange(key)
+    // Define the setUnknownProperty function on the source property so that we can
+    // monitor for the addition of new properties and spread them onto the local object
+    defineProperty(get(sourceObject, sourceProperty), 'setUnknownProperty', undefined,
+      function (key, value) {
+        // Set the property to the given value (the expected normal behavior)
+        this[key] = value
+
+        // For each listening target object (registered via spread options)
+        // spread the new property onto the target object
+        this._spreadListeners.forEach(listener => {
+          if (typeOf(value) === 'function') {
+            listener.targetObject.set(key, value)
+          } else {
+            defineProperty(listener.targetObject, key,
+              computed.readOnly(`${listener.targetProperty}.${key}`)
+            )
           }
-        )
+        })
+
+        // Notify all downstream listeners that the property has changed.
+        // This triggers the first observation of the property for the newly
+        // defined computed property on the target object(s)
+        sourceObject.get(sourceProperty).notifyPropertyChange(key)
+      }
+    )
+  },
+
+  /**
+   * Create local properties for each property in the spread hash.
+   * Functions are set directly against the local object, while all
+   * other properties are readOnly computed properties to retain
+   * observer behavior
+   *
+   * Note: We're currently using the private Ember defineProperty function
+   * which is required to establish observer chains (accept computed properties)
+   *
+   * @param {string} spreadProperty - the name of the local property containing the hash
+   * @param {object} spreadableHash - the hash to spread
+   */
+  _defineSpreadProperties (spreadProperty, spreadableHash) {
+    assert(`${spreadProperty} requires an Ember object or primitive object`,
+      typeOf(spreadableHash) === 'instance' ||
+      typeOf(spreadableHash) === 'object'
+    )
```

Comment:
Sure, I'm open to a convention here - I think the second style is probably the better version, if a little less readable.  I'll swap to that for now and we can see if other people agree.